src/Features/CSharp/Portable/AddBraces/CSharpAddBracesCodeFixProvider.cs

```diff
@@ -1,17 +1,37 @@
-﻿using System.Composition;
+﻿using System;
+using System.Collections.Immutable;
+using System.Composition;
 using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.CodeAnalysis.AddBraces;
+using Microsoft.CodeAnalysis.CodeActions;
 using Microsoft.CodeAnalysis.CodeFixes;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.AddBraces
 {
     [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.AddBraces), Shared]
     [ExtensionOrder(After = PredefinedCodeFixProviderNames.AddAwait)]
-    internal class CSharpAddBracesCodeFixProvider : AbstractAddBracesCodeFixProvider
+    internal class CSharpAddBracesCodeFixProvider : CodeFixProvider
     {
-        protected override async Task<Document> AddBracesAsync(CodeFixContext context)
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IDEDiagnosticIds.AddBracesDiagnosticId);
+
+        public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(
+                new MyCodeAction(
+                    FeaturesResources.AddBraces,
+                    async c => await AddBracesAsync(context).ConfigureAwait(false)),
+                context.Diagnostics);
+
+            return SpecializedTasks.EmptyTask;
+        }
+
+        protected async Task<Document> AddBracesAsync(CodeFixContext context)
         {
             var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
```

Comment:
So i don't want unnecessary *work* to be done in the registration method.  i.e. don't do things that are *expensive* and not needed until the fix is actually applied.

What is needed in the registration method is:

1. Any work necessary to determine which code fixes to provide.  For example, if multiple fixes may be given, but those fixes are conditioned on tests you need to perform.
2. Extracting out appropriate data *cheaply* from teh context menu to pass to the fixes.

The reason i don't like passing the context around is that there's a good chance the wrong cancellation token will be used.  (even i've made that mistake).

So the ideal code here for me would be:

```c#
c => await AddBracesAsync(context.Whatever, context.Whatever, ..., c);
```

Then AddBracesAsync only has what it needs, and only sees one potential cancellation token.